### PR TITLE
Nrfx 4720 recalculate temperature when ntc changed

### PR DESCRIPTION
**Pull request by magp-nordic**
Recalculation of NTC resistance after beta changes for temperatures set beforehand. Change of NTC type does not affect the code value written to the register (equation from Figure 23 from IPS v 0.53).
Die temperatures do not depend on NTC type or beta value.
Recalculation is done by reading the old resistance value, calculating it to the temperature value, and then recalculating resistance for that temperature. Although it requires additional computing, it does not increase the execution time of a command in a noticeable way.

test:PR-177

### CODE REVIEW

**Comment by nordic-segl** [comment]
Tests for the added functionality in the shell sample are implemented in this PR:
https://projecttools.nordicsemi.no/bitbucket/projects/NCS-TEST/repos/test-sdk-npmx/pull-requests/177/overview

**Comment by nordic-segl** [comment]
Waiting for results of integration CI that includes new tests
https://jenkins-ncs.nordicsemi.no/job/HCS/job/NPMX/job/npmx_zephyr_integration/job/PR-177/

File: drivers/npmx/npmx_shell.c [hunk 1]
```diff
@@ -2940,6 +2994,44 @@ static int adc_ntc_set(const struct shell *shell, size_t argc, char **argv,
...
+			shell_error(shell,
+				    "Error: unable to recalculate NTC for cold temperature.");
+			return 0;
+		}
+
+		if (!adc_ntc_temperatures_recalculate(shell, charger_instance, &ntc_config_old,
+						      &ntc_config, npmx_charger_cool_resistance_get,
+						      npmx_charger_cool_resistance_set)) {
+			shell_error(shell,
+				    "Error: unable to recalculate NTC for cold temperature.");
```
Comment (nordic-segl) [inline drivers/npmx/npmx_shell.c:3016]: for cool

**Comment by jaz1-nordic** [comment]
I see that you are changing these functions, so maybe it is worth considering in scope of this PR this task:
https://nordicsemi.atlassian.net/browse/NRFX-4708

**Review by jaz1-nordic — DISMISSED** [review]

**Comment by magp-nordic** [comment]
> I see that you are changing these functions, so maybe it is worth considering in scope of this PR this task: https://nordicsemi.atlassian.net/browse/NRFX-4708

I thought about it, but from what I can see, npmx-zephyr takes npmx from the public repository, so I would have to make a release, to include those changes here. I think it would be better to create a separate PR in npmx-zephyr with replacing calculations with npmx function when npmx 1.0 is released.

File: drivers/npmx/npmx_shell.c [hunk 2]
```diff
@@ -859,6 +859,27 @@ static int cmd_ntc_resistance_hot_get(const struct shell *shell, size_t argc, ch
 	return cmd_ntc_resistance_get(shell, argc, argv, npmx_charger_hot_resistance_get);
 }
 
+static bool ntc_temperature_from_resistance(const struct shell *shell,
+					    npmx_adc_ntc_config_t *ntc_config, uint32_t resistance,
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:863]:
```suggestion
					    const npmx_adc_ntc_config_t *ntc_config, uint32_t resistance,
```

File: drivers/npmx/npmx_shell.c [hunk 3]
```diff
@@ -859,6 +859,27 @@ static int cmd_ntc_resistance_hot_get(const struct shell *shell, size_t argc, ch
 	return cmd_ntc_resistance_get(shell, argc, argv, npmx_charger_hot_resistance_get);
 }
 
+static bool ntc_temperature_from_resistance(const struct shell *shell,
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:862]:
```suggestion
static bool ntc_resistance_to_temperature(const struct shell *shell,
```

File: drivers/npmx/npmx_shell.c [hunk 4]
```diff
@@ -882,30 +903,40 @@ static int cmd_ntc_temperature_get(const struct shell *shell, size_t argc, char
...
 
 	if (check_error_code(shell, err_code)) {
-		float numerator = THERMISTOR_NOMINAL_TEMPERATURE * (float)ntc_config.beta;
-		float denominator = (THERMISTOR_NOMINAL_TEMPERATURE *
-				     log((float)resistance / (float)ntc_nominal_resistance)) +
-				    (float)ntc_config.beta;
-		float temperature = round((numerator / denominator) - ABSOLUTE_ZERO_DIFFERENCE);
-		shell_print(shell, "Value: %d *C.", (int)temperature);
+		int16_t temperature;
+		ntc_temperature_from_resistance(shell, &ntc_config, resistance, &temperature);
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:912]: There should be an error checker.

File: drivers/npmx/npmx_shell.c [hunk 5]
```diff
@@ -882,30 +903,40 @@ static int cmd_ntc_temperature_get(const struct shell *shell, size_t argc, char
...
+		shell_print(shell, "Value: %d *C.", temperature);
 	} else {
 		shell_error(shell, "Error: unable to read NTC resistance value.");
 	}
 
 	return 0;
 }
 
+static bool ntc_resistance_from_temperature(const struct shell *shell, int16_t temperature,
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:921]:
```suggestion
static bool ntc_temperature_to_resistance(const struct shell *shell, int16_t temperature,
```

File: drivers/npmx/npmx_shell.c [hunk 6]
```diff
@@ -882,30 +903,40 @@ static int cmd_ntc_temperature_get(const struct shell *shell, size_t argc, char
...
+					    npmx_adc_ntc_config_t *ntc_config, uint32_t *resistance)
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:922]:
```suggestion
					    const npmx_adc_ntc_config_t *ntc_config, uint32_t *resistance)
```

File: drivers/npmx/npmx_shell.c [hunk 7]
```diff
@@ -2819,6 +2844,32 @@ static int adc_ntc_get(const struct shell *shell, size_t argc, char **argv,
...
+	uint32_t resistance;
+
+	npmx_error_t err_code = get_func(charger_instance, &resistance);
+	if (!check_error_code(shell, err_code)) {
+		shell_error(shell, "Error: unable to get old resistance value.");
+		return false;
+	}
+	ntc_temperature_from_resistance(shell, ntc_config_old, resistance, &temperature);
+	ntc_resistance_from_temperature(shell, temperature, ntc_config_new, &resistance);
+	err_code = set_func(charger_instance, resistance);
```
Comment (wilu-nordic) [inline drivers/npmx/npmx_shell.c:2863]: Please add error checkers.

File: drivers/npmx/npmx_shell.c [hunk 8]
```diff
@@ -2819,6 +2844,32 @@ static int adc_ntc_get(const struct shell *shell, size_t argc, char **argv,
 	return 0;
 }
 
+static bool adc_ntc_temperatures_recalculate(
+	const struct shell *shell, npmx_charger_t *charger_instance,
+	npmx_adc_ntc_config_t *ntc_config_old, npmx_adc_ntc_config_t *ntc_config_new,
+	npmx_error_t (*get_func)(npmx_charger_t const *p_instance, uint32_t *resistance),
+	npmx_error_t (*set_func)(npmx_charger_t const *p_instance, uint32_t resistance))
```
Comment (masz-nordic) [inline drivers/npmx/npmx_shell.c:2851]: A helper struct could be created to aggregate parameters (except for `*shell`).

**Comment by masz-nordic** [comment]
Commit title format should be changed to Zephyr style.

File: drivers/npmx/npmx_shell.c (hunk 2 above)
Comment (magp-nordic) [reply drivers/npmx/npmx_shell.c]: Added after `npmx_adc_ntc_config_t`

File: drivers/npmx/npmx_shell.c (hunk 6 above)
Comment (magp-nordic) [reply drivers/npmx/npmx_shell.c]: Same as above

File: drivers/npmx/npmx_shell.c [hunk 9]
```diff
@@ -2993,6 +3035,35 @@ static int adc_ntc_get(const struct shell *shell, size_t argc, char **argv,
...
+	npmx_error_t err_code = data->get_resistance_func(data->charger_instance, &resistance);
+	if (!check_error_code(shell, err_code)) {
+		shell_error(shell, "Error: unable to get old resistance value.");
+		return false;
+	}
+	if (!ntc_resistance_to_temperature(shell, data->ntc_config_old, resistance, &temperature)) {
+		shell_error(shell, "Error: unable convert old resistance to temperature.");
+		return false;
+	}
+	if (!ntc_temperature_to_resistance(shell, temperature, data->ntc_config_new, &resistance)) {
```
Comment (masz-nordic) [inline drivers/npmx/npmx_shell.c:3053]: Can the pointer to `data` be second parameter in `ntc_temperature_to_resistance()`?
It would be more readable.

**Review by masz-nordic — APPROVED** [review]

**Review by wilu-nordic — APPROVED** [review]

**Pull request merged**